File: deep-neuro/patch_files/rivercrossing/tf_rivercrossing.cpp
/*
Copyright (c) 2018 Keele Technologies, Inc.
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#include <iostream>
#include <string>
#include "tensorflow/core/framework/op_kernel.h"
#include "tensorflow/core/framework/op.h"
#include "tensorflow/core/framework/shape_inference.h"
#include "tensorflow/core/framework/resource_mgr.h"
#include "tensorflow/core/framework/resource_op_kernel.h"
#include "tensorflow/core/lib/core/blocking_counter.h"
#include "tensorflow/core/lib/core/threadpool.h"
#include "tensorflow/core/platform/mutex.h"
#include "tf_env.h"
#include "rivercrossing.h"

#ifdef __USE_SDL
  #include <SDL.h>
#endif

using namespace tensorflow;
using namespace std;

class RivercrossingEnvironment : public Environment<float>, public StepInterface<int>
{
    public:
        int tensor_layers;
        string input_tensor;
        RivercrossingEnvironment(string tensor, int layers, int batch_size) //Complete
        {

            m_pInterfaces = new river::Environment[batch_size];
            m_numSteps.resize(batch_size, 0);
            input_tensor = tensor;
            tensor_layers = layers;

        }

        virtual ~RivercrossingEnvironment() { //Complete
            // cout << "River Enviroment Deleted" << endl;
            delete[] m_pInterfaces;
        }

        TensorShape get_observation_shape() override //Complete
        {
            // cout << "River Enviroment returned Observation Shape" << endl;

            return TensorShape({
                static_cast<int>(world::x),
                static_cast<int>(world::y),
                static_cast<int>(tensor_layers)
        });
            //InvalidArgumentError (see above for traceback): Input to reshape is a tensor with THIS values, but the requested shape requires a multiple of 1200
        }

        void get_observation(float *data, int idx) override //Complete
        {
            // cout << "River Enviroment returned Observation Inputs" << endl;
            assert(idx < m_numSteps.size());
            m_pInterfaces[idx].generate_neural_inputs(data, input_tensor);
        }

        TensorShape get_action_shape() override { //Complete
            return TensorShape();
        }

        float step(int idx, const int* action) override { //Complete
            // cout << "River Enviroment stepped" << endl;
            assert(idx < m_numSteps.size());
            m_numSteps[idx] = m_pInterfaces[idx].Update(action);

            if (is_done(idx))
            {
                return m_pInterfaces[idx].return_fitness();
            }
            return 0.0f;
        }

        bool is_done(int idx) override { //Complete
            // cout << "River Enviroment is been checked if done" << endl;
            return m_pInterfaces[idx].has_failed() || m_numSteps[idx] >= 150;
        }

        void reset(int i, int numNoops=0, int maxFrames=100000) override {
            // cout << "River Enviroment has been reset" << endl;
            m_pInterfaces[i].reset();
            m_numSteps[i] = 0;
        }

        string DebugString() override { return "RivercrossingEnvironment"; } //Complete
     private:
        river::Environment* m_pInterfaces;
        std::vector<int> m_numSteps;
};

class RivercrossingMakeOp : public EnvironmentMakeOp {
    public:
    explicit RivercrossingMakeOp(OpKernelConstruction* context) : EnvironmentMakeOp(context) {
        OP_REQUIRES_OK(context, context->GetAttr("input_tensor", &input_tensor));
        OP_REQUIRES_OK(context, context->GetAttr("num_layers", &num_layers));
    }

 private:
    virtual Status CreateResource(OpKernelContext* context, BaseEnvironment** ret) EXCLUSIVE_LOCKS_REQUIRED(mu_) {
        RivercrossingEnvironment* env = new RivercrossingEnvironment(input_tensor, num_layers, batch_size);
        if (env == nullptr)
            return errors::ResourceExhausted("Failed to allocate");
        *ret = env;

        const auto thread_pool = context->device()->tensorflow_cpu_worker_threads();
        const int num_threads = std::min(program::threads, batch_size);
        auto f = [&](int thread_id) {
            for(int b =thread_id; b < batch_size;b+=num_threads)
            {
                env->reset(b);
            }
        };

        BlockingCounter counter(num_threads-1);
        for (int i = 1; i < num_threads; ++i) {
            thread_pool->workers->Schedule([&, i]() {
                f(i);
                counter.DecrementCount();
            });
        }
        f(0); //Running scheduled thread
        counter.Wait();
        return Status::OK();
    }
    std::string input_tensor;
    int num_layers;
};

REGISTER_OP("RivercrossingMake")
    .Attr("input_tensor: string")
    .Attr("num_layers: int")
    .Attr("batch_size: int")
    .Attr("container: string = ''")
    .Attr("shared_name: string = ''")
    .Output("handle: resource")
    .SetIsStateful()
    .SetShapeFn(shape_inference::ScalarShape);

REGISTER_KERNEL_BUILDER(Name("RivercrossingMake").Device(DEVICE_CPU), RivercrossingMakeOp);

File: deep-neuro/patch_files/rivercrossing/tf_rivercrossing.py
__copyright__ = """
Copyright (c) 2018 Uber Technologies, Inc.
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
"""
import tensorflow as tf
from gym_tensorflow.tf_env import TensorFlowEnv, gym_tensorflow_module
import numpy as np

types_of_tensor = ['RGB','objects','greyscale']
tensor_layer_size = [3,6,1]
RC_size_x = 20
RC_size_y = 20

class RivercrossingEnv(TensorFlowEnv):
    def __init__(self, input_tensor, batch_size, name=None, env_size=(RC_size_x, RC_size_y) ):
        assert input_tensor in types_of_tensor, "{} is not part of the available River Crossing Tensors".format(input_tensor)

        self.input_tensor = input_tensor
        self.batch_size = batch_size
        self.env_size = env_size

        # choose the apprioate number of layers 
        for i in range(len(types_of_tensor)):
             if (self.input_tensor ==    types_of_tensor[i]):
                self.num_stacked_frames = tensor_layer_size[i]

        with tf.variable_scope(name, default_name='RiverInstance'):
            self.instances = gym_tensorflow_module.rivercrossing_make(input_tensor=input_tensor, num_layers=self.num_stacked_frames, batch_size=batch_size)
        self.obs_variable = tf.Variable(tf.zeros(shape=self.observation_space, dtype=tf.int32), trainable=False)

    @property
    def observation_space(self):
        return (self.batch_size, ) + self.env_size + (self.num_stacked_frames,)

    @property
    def env_default_timestep_cutoff(self):
        return 20000000

    """
    Amount of output from network
    """
    @property 
    def action_space(self):
        return 18

    @property
    def discrete_action(self):
        return True

    def step(self, action, indices=None, name=None):
        if indices is None:
            indices = np.arange(self.batch_size)
        with tf.variable_scope(name, default_name='RivercrossingStep'):
            rew, done = gym_tensorflow_module.environment_step(self.instances, indices, action)
            return rew, done

    def reset(self, indices=None, max_frames=None, name=None):
        '''Resets River Crossing instances with a random noop start (1-30) and set the maximum number of frames for the episode (default 100,000 * frameskip)
        '''
        if indices is None:
            indices = np.arange(self.batch_size)

        with tf.variable_scope(name, default_name='RivercrossingReset'):
            noops = tf.random_uniform(tf.shape(indices), minval=1, maxval=31, dtype=tf.int32)
            if max_frames is None:
                max_frames = self.env_default_timestep_cutoff
            return gym_tensorflow_module.environment_reset(self.instances, indices, noops=noops, max_frames=max_frames)

    def observation(self, indices=None, name=None):
        if indices is None:
            indices = np.arange(self.batch_size)
        with tf.variable_scope(name, default_name='RivercrossingObservation'):
            with tf.device('/cpu:0'):
                obs = gym_tensorflow_module.environment_observation(self.instances, indices, T=tf.float32)
            obs.set_shape((None,) + self.env_size + (self.num_stacked_frames,))
        return obs


    def close(self):
        pass